lib/src/token.cc

```diff
@@ -0,0 +1,48 @@
+#include <internal/token.hpp>
+#include <internal/simple_config_origin.hpp>
+
+using namespace std;
+
+namespace hocon {
+
+    unsupported_exception::unsupported_exception(string const& message) :
+        runtime_error(message) { }
+
+    token::token(token_type type, shared_ptr<simple_config_origin> origin, string token_text, string debug_string) :
+        _token_type(type), _origin(move(origin)), _token_text(move(token_text)),
+        _debug_string(move(debug_string)) { }
+
+
+    token_type token::get_token_type() const {
+        return _token_type;
+    }
+
+    string token::token_text() const {
+        return _token_text;
+    }
+
+    string token::to_string() const {
+        return _debug_string;
+    }
+
+    int token::line_number() const {
+        if (_origin) {
+            return _origin->line_number();
+        } else {
+            return -1;
+        }
+    }
+
+    shared_ptr<simple_config_origin> token::origin() const {
```

Comment:
We should return this by const&. We're granting access to an object owned by token, but we don't want to make a copy of the shared_ptr (which is more expensive than returning a reference) unless someone else also wants to take ownership of it. So if we return a const& here, others can use it, and take by value when they need to copy it.